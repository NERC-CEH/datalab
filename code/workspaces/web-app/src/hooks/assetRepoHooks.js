import useShallowSelector from './useShallowSelector';
import assetRepoSelectors from '../selectors/assetRepoSelectors';
import sortByName from '../components/common/sortByName';
import { PUBLIC } from '../components/assetRepo/assetVisibilities';

const useAssetRepoUnsorted = () => useShallowSelector(assetRepoSelectors.assetRepo);

export const useAssetRepo = () => {
  const assetRepo = useAssetRepoUnsorted();
  return {
    ...assetRepo,
    value:
    {
      ...assetRepo.value,
      assets: sortByName(assetRepo.value.assets),
    },
  };
};

const assetVisible = (projectKey, asset) => asset.fileLocation
  && (asset.visible === PUBLIC
  || (asset.projects && asset.projects.map(project => project.key).includes(projectKey)));

export const useVisibleAssets = (projectKey) => {
  const assetRepo = useAssetRepoUnsorted();
  const visibleAssets = assetRepo.value.assets
    ? assetRepo.value.assets.filter(asset => assetVisible(projectKey, asset))
    : [];
  return {
    ...assetRepo,
    value:
    {
      ...assetRepo.value,
      assets: sortByName(visibleAssets),
    },
  };
};
