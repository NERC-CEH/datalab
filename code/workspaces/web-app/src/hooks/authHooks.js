import useShallowSelector from './useShallowSelector';
import authSelectors from '../selectors/authSelectors';

const { currentUserId, currentUserPermissions, currentUserTokens } = authSelectors;

export const useCurrentUserId = () => useShallowSelector(currentUserId);
export const useCurrentUserPermissions = () => useShallowSelector(currentUserPermissions);
export const useCurrentUserTokens = () => useShallowSelector(currentUserTokens);
