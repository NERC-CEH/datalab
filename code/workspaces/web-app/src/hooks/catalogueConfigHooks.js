import useShallowSelector from './useShallowSelector';
import catalogueConfigSelectors from '../selectors/catalogueConfigSelectors';

export const useCatalogueConfig = () => useShallowSelector(catalogueConfigSelectors.catalogueConfig);
export const useCatalogueAvailable = () => useShallowSelector(catalogueConfigSelectors.catalogueAvailable);
