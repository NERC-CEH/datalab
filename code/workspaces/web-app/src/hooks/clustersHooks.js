import useShallowSelector from './useShallowSelector';
import clustersSelectors from '../selectors/clustersSelectors';

// eslint-disable-next-line import/prefer-default-export
export const useClusters = () => useShallowSelector(clustersSelectors.clusters);
export const useClustersByType = (type) => {
  const clusters = useClusters();

  if (!type) {
    return clusters;
  }

  return {
    ...clusters,
    value: clusters.value.filter(cluster => cluster.type === type),
  };
};
