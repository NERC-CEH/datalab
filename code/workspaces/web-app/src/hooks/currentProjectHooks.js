import useShallowSelector from './useShallowSelector';
import currentProjectSelectors from '../selectors/currentProjectSelectors';

const { currentProject, currentProjectKey } = currentProjectSelectors;

export const useCurrentProject = () => useShallowSelector(currentProject);
export const useCurrentProjectKey = () => useShallowSelector(currentProjectKey);
