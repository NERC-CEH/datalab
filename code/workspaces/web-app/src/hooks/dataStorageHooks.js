import useShallowSelector from './useShallowSelector';
import dataStorageSelectors from '../selectors/dataStorageSelectors';

// eslint-disable-next-line import/prefer-default-export
export const useDataStorageArray = () => useShallowSelector(dataStorageSelectors.dataStorageArray);
export const useDataStorageForUserInProject = (userId, projectKey) => {
  const dataStores = useDataStorageArray();
  return {
    ...dataStores,
    value: dataStores.value.filter(store => store.projectKey === projectKey && store.users.includes(userId)),
  };
};
