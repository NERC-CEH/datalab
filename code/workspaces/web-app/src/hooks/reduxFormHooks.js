import { useSelector } from 'react-redux';
import reduxFormSelectors from '../selectors/reduxFormSelectors';

// eslint-disable-next-line import/prefer-default-export
export const useReduxFormValue = (formName, fieldName) => useSelector(
  reduxFormSelectors.reduxFormValue(formName, fieldName),
);
