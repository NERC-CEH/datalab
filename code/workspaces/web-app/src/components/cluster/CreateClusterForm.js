import { Field, reduxForm } from 'redux-form';
import PropTypes from 'prop-types';
import React from 'react';
import { CreateFormControls, renderSelectField, renderTextField } from '../common/form/controls';
import { useReduxFormValue } from '../../hooks/reduxFormHooks';
import { syncValidate } from './createClusterValidator';

const commonFieldProps = {
  component: renderTextField,
  InputLabelProps: { shrink: true },
};

const createHelperText = ({ lowerLimit, default: defaultValue, upperLimit }) => (
  `Value range: min=${lowerLimit} max=${upperLimit} default=${defaultValue}`
);

const noneOption = { text: 'None', value: '--none--' };
export const removeNoneOptions = values => Object.entries(values).reduce(
  (filtered, [key, value]) => {
    if (value !== noneOption.value) return { ...filtered, [key]: value };
    return filtered;
  },
  {},
);

export const CreateClusterFormContent = ({
  handleSubmit, form: formName, onSubmit, cancel, submitting, dataStorageOptions, clusterMaxWorkers, workerMaxMemory, workerMaxCpu,
}) => {
  const VOLUME_MOUNT_FIELD_NAME = 'volumeMount';
  const volumeMountValue = useReduxFormValue(formName, VOLUME_MOUNT_FIELD_NAME);

  return (
    <form onSubmit={handleSubmit(values => onSubmit(removeNoneOptions(values)))}>
      <Field
        {...commonFieldProps}
        name="displayName"
        label="Display Name"
      />
      <Field
        {...commonFieldProps}
        name="name"
        label="Internal Name"
      />
      <Field
        {...commonFieldProps}
        name={VOLUME_MOUNT_FIELD_NAME}
        label="Data Store to Mount"
        component={renderSelectField}
        options={[noneOption, ...dataStorageOptions]}
      />
      {volumeMountValue && volumeMountValue !== noneOption.value
        && <Field
          {...commonFieldProps}
          name="condaPath"
          label="Path to Conda environment (optional)"
          helperText="The file path to the conda environment on the selected Data Store to be used by the workers."
        />
      }
      <Field
        {...commonFieldProps}
        name="maxWorkers"
        label="Maximum number of workers"
        type="number"
        InputProps={{
          inputProps: {
            min: clusterMaxWorkers.lowerLimit,
            max: clusterMaxWorkers.upperLimit,
            step: 1,
          },
        }}
        parse={value => Number(value)}
        helperText={createHelperText(clusterMaxWorkers)}
      />
      <Field
        {...commonFieldProps}
        name="maxWorkerMemoryGb"
        label="Maximum worker memory (GB)"
        type="number"
        InputProps={{
          inputProps: {
            min: workerMaxMemory.lowerLimit,
            max: workerMaxMemory.upperLimit,
            step: 0.5,
          },
        }}
        parse={value => Number(value)}
        helperText={createHelperText(workerMaxMemory)}
      />
      <Field
        {...commonFieldProps}
        name="maxWorkerCpu"
        label="Maximum worker CPU use"
        type="number"
        InputProps={{
          inputProps: {
            min: workerMaxCpu.lowerLimit,
            max: workerMaxCpu.upperLimit,
            step: 0.5,
          },
        }}
        parse={value => Number(value)}
        helperText={createHelperText(workerMaxCpu)}
      />
      <CreateFormControls onCancel={cancel} submitting={submitting}/>
    </form>
  );
};

const CreateClusterReduxForm = reduxForm({
  validate: syncValidate,
  destroyOnUnmount: false,
  enableReinitialize: true,
})(CreateClusterFormContent);

const CreateClusterForm = ({ formName, ...otherProps }) => <CreateClusterReduxForm form={formName} {...otherProps} />;

export { CreateClusterFormContent as PureCreateClusterForm };
export default CreateClusterForm;

const dropDownOptionType = PropTypes.shape({
  text: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
});

const dropDownOptionArrayType = PropTypes.arrayOf(dropDownOptionType);

const inputConstraintPropTypes = PropTypes.shape({
  lowerLimit: PropTypes.number.isRequired,
  upperLimit: PropTypes.number.isRequired,
  default: PropTypes.number.isRequired,
});

CreateClusterForm.propTypes = {
  formName: PropTypes.string.isRequired,
  onSubmit: PropTypes.func.isRequired,
  cancel: PropTypes.func.isRequired,
  dataStorageOptions: dropDownOptionArrayType.isRequired,
  clusterMaxWorkers: inputConstraintPropTypes.isRequired,
  workerMaxMemory: inputConstraintPropTypes.isRequired,
  workerMaxCpu: inputConstraintPropTypes.isRequired,
};
