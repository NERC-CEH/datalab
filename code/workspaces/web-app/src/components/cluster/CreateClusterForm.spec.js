import React from 'react';
import { shallow } from 'enzyme';
import { PureCreateClusterForm, removeNoneOptions } from './CreateClusterForm';
import { useReduxFormValue } from '../../hooks/reduxFormHooks';

jest.mock('../../hooks/reduxFormHooks');

const getProps = () => ({
  handleSubmit: jest.fn().mockName('handleSubmit'),
  onSubmit: jest.fn().mockName('onSubmit'),
  cancel: jest.fn().mockName('cancel'),
  submitting: false,
  dataStorageOptions: [{ text: 'Test Store', value: 'teststore' }],
  clusterMaxWorkers: { lowerLimit: 1, default: 4, upperLimit: 8 },
  workerMaxMemory: { lowerLimit: 0.5, default: 4, upperLimit: 8 },
  workerMaxCpu: { lowerLimit: 0.5, default: 0.5, upperLimit: 2 },
  condaRequired: false,
});

describe('CreateClusterForm', () => {
  describe('it renders to match snapshot', () => {
    it('without condaPath field when volumeMount field has no value', () => {
      useReduxFormValue.mockReturnValueOnce(null);
      const props = getProps();
      const render = shallow(<PureCreateClusterForm {...props} />);
      expect(render).toMatchSnapshot();
    });

    it('without condaPath field when volumeMount field is set as None option', () => {
      useReduxFormValue.mockReturnValueOnce('--none--');
      const props = getProps();
      const render = shallow(<PureCreateClusterForm {...props} />);
      expect(render).toMatchSnapshot();
    });

    it('with condaPath field when volumeMount field is set', () => {
      useReduxFormValue.mockReturnValueOnce('teststore');
      const props = getProps();
      const render = shallow(<PureCreateClusterForm {...props} />);
      expect(render).toMatchSnapshot();
    });

    it('when conda is required and volumeMount field is set', () => {
      useReduxFormValue.mockReturnValueOnce('teststore');
      const props = {
        ...getProps(),
        condaRequired: true,
      };
      const render = shallow(<PureCreateClusterForm {...props} />);
      expect(render).toMatchSnapshot();
    });
  });
});

describe('removeNoneOptions', () => {
  it('removes all items in provided values object that contain the none option value', () => {
    const noneValue = '--none--';
    const values = {
      name: 'testcluster',
      displayName: 'Test Cluster',
      volumeMount: noneValue,
      maxWorkers: 4,
    };
    const result = removeNoneOptions(values);
    expect(result.name).toEqual(values.name);
    expect(result.displayName).toEqual(values.displayName);
    expect(result.maxWorkers).toEqual(values.maxWorkers);
    expect(result.volumeMount).toBeUndefined();
  });
});
