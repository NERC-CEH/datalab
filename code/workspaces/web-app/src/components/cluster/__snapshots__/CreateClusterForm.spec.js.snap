// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateClusterForm it renders to match snapshot when conda is required and volumeMount field is set 1`] = `
<div>
  <form>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"displayName","label":"Display Name"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"name","label":"Internal Name"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"volumeMount","label":"Data Store to Mount","options":[{"text":"Test Store","value":"teststore"}]}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"condaPath","label":"Path to Conda environment","helperText":"The file path to the conda environment on the selected Data Store to be used by the workers.","placeholder":"/data/conda/conda-env"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkers","label":"Maximum number of workers","type":"number","InputProps":{"inputProps":{"min":1,"max":8,"step":1}},"helperText":"Value range: min=1 max=8 default=4"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkerMemoryGb","label":"Maximum worker memory (GB)","type":"number","InputProps":{"inputProps":{"min":0.5,"max":8,"step":0.5}},"helperText":"Value range: min=0.5 max=8 default=4"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkerCpu","label":"Maximum worker CPU use","type":"number","InputProps":{"inputProps":{"min":0.5,"max":2,"step":0.5}},"helperText":"Value range: min=0.5 max=2 default=0.5"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"assets","label":"Assets"}
    </div>
    <div
      class="makeStyles-fromControlContainer-7"
    >
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-o0h1hg-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Create
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-4eo762-MuiButtonBase-root-MuiButton-root"
        style="margin-left: 5px;"
        tabindex="0"
        type="button"
      >
        Cancel
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`CreateClusterForm it renders to match snapshot with condaPath field when volumeMount field is set 1`] = `
<div>
  <form>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"displayName","label":"Display Name"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"name","label":"Internal Name"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"volumeMount","label":"Data Store to Mount (optional)","options":[{"text":"None","value":"--none--"},{"text":"Test Store","value":"teststore"}]}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"condaPath","label":"Path to Conda environment (optional)","helperText":"The file path to the conda environment on the selected Data Store to be used by the workers.","placeholder":"/data/conda/conda-env"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkers","label":"Maximum number of workers","type":"number","InputProps":{"inputProps":{"min":1,"max":8,"step":1}},"helperText":"Value range: min=1 max=8 default=4"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkerMemoryGb","label":"Maximum worker memory (GB)","type":"number","InputProps":{"inputProps":{"min":0.5,"max":8,"step":0.5}},"helperText":"Value range: min=0.5 max=8 default=4"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkerCpu","label":"Maximum worker CPU use","type":"number","InputProps":{"inputProps":{"min":0.5,"max":2,"step":0.5}},"helperText":"Value range: min=0.5 max=2 default=0.5"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"assets","label":"Assets"}
    </div>
    <div
      class="makeStyles-fromControlContainer-5"
    >
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-o0h1hg-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Create
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-4eo762-MuiButtonBase-root-MuiButton-root"
        style="margin-left: 5px;"
        tabindex="0"
        type="button"
      >
        Cancel
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`CreateClusterForm it renders to match snapshot without condaPath field when volumeMount field has no value 1`] = `
<div>
  <form>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"displayName","label":"Display Name"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"name","label":"Internal Name"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"volumeMount","label":"Data Store to Mount (optional)","options":[{"text":"None","value":"--none--"},{"text":"Test Store","value":"teststore"}]}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkers","label":"Maximum number of workers","type":"number","InputProps":{"inputProps":{"min":1,"max":8,"step":1}},"helperText":"Value range: min=1 max=8 default=4"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkerMemoryGb","label":"Maximum worker memory (GB)","type":"number","InputProps":{"inputProps":{"min":0.5,"max":8,"step":0.5}},"helperText":"Value range: min=0.5 max=8 default=4"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkerCpu","label":"Maximum worker CPU use","type":"number","InputProps":{"inputProps":{"min":0.5,"max":2,"step":0.5}},"helperText":"Value range: min=0.5 max=2 default=0.5"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"assets","label":"Assets"}
    </div>
    <div
      class="makeStyles-fromControlContainer-1"
    >
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-o0h1hg-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Create
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-4eo762-MuiButtonBase-root-MuiButton-root"
        style="margin-left: 5px;"
        tabindex="0"
        type="button"
      >
        Cancel
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`CreateClusterForm it renders to match snapshot without condaPath field when volumeMount field is set as None option 1`] = `
<div>
  <form>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"displayName","label":"Display Name"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"name","label":"Internal Name"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"volumeMount","label":"Data Store to Mount (optional)","options":[{"text":"None","value":"--none--"},{"text":"Test Store","value":"teststore"}]}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkers","label":"Maximum number of workers","type":"number","InputProps":{"inputProps":{"min":1,"max":8,"step":1}},"helperText":"Value range: min=1 max=8 default=4"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkerMemoryGb","label":"Maximum worker memory (GB)","type":"number","InputProps":{"inputProps":{"min":0.5,"max":8,"step":0.5}},"helperText":"Value range: min=0.5 max=8 default=4"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"maxWorkerCpu","label":"Maximum worker CPU use","type":"number","InputProps":{"inputProps":{"min":0.5,"max":2,"step":0.5}},"helperText":"Value range: min=0.5 max=2 default=0.5"}
    </div>
    <div>
      Field mock 
      {"InputLabelProps":{"shrink":true},"name":"assets","label":"Assets"}
    </div>
    <div
      class="makeStyles-fromControlContainer-3"
    >
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-o0h1hg-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Create
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-4eo762-MuiButtonBase-root-MuiButton-root"
        style="margin-left: 5px;"
        tabindex="0"
        type="button"
      >
        Cancel
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;
